fix(compose, dev): improve `--graph-ref` option

glasser opened the pull request:
Make `rover supergraph compose --graph-ref REF --config CONFIG` work as documented (and don't remove `routing_url`s fetched from GraphOS if they're missing from the local file).

Make `rover supergraph compose --graph-ref REF` and `rover dev --graph-ref REF` use the graph ref's federation version.

Fix docs.

(Individual commits have more detailed messages.)

glasser committed:
compose and dev: read federation version when `--graph-ref` is specified

The `rover dev` and `rover supergraph compose` commands take a
`--graph-ref` option (since v0.25.0 and v0.26.0 respectively) which
loads subgraph definitions from GraphOS. These commands also take
optional `--federation-version` arguments to let you specify the
federation version, and you can also specify the federation version
inside a supergraph.yaml file named with the `rover dev
--supergraph-config`/`rover supergraph compose --config` option.

Previously, if you did not specify a federation version via either of
these mechanisms, the command would default to using the latest Fed v2
version.

However, Studio has a specified federation version for each graph ref,
which would be a better default for the case that the user does not
specify the version directly.

This change reads the graph's federation version from GraphOS in the
same query that reads subgraph information, and uses that as the default
federation version if a version is not specified via flag (priority 1)
or config file (priority 2).

glasser committed:
compose: allow specifying `--graph-ref` and `--config` at the same time

The actual implementation of the command already understands how to
combine these flags, but their declaration did not let you do it.

glasser committed:
docs: document `rover supergraph compose` correctly

There were a few typos in #2037 (and the override feature didn't work
until this PR anyway).

aaronArinder commented on crates/rover-client/src/operations/subgraph/fetch_all/types.rs:
```diff
@@ -72,3 +80,31 @@ impl
+    fn from(
+        value: subgraph_fetch_all_query::SubgraphFetchAllQueryVariantOnGraphVariantSourceVariantLatestLaunch,
+    ) -> Self {
+        if let subgraph_fetch_all_query::SubgraphFetchAllQueryVariantOnGraphVariantSourceVariantLatestLaunchBuildInput::CompositionBuildInput(composition_build_input) = value.build_input {
```
this is a bit ugly; can we slim down the reference to CompositionBuildInput by importing it and the others?

glasser replied:
Unfortunately there are two different CompositionBuildInputs in the two functions, so I'm not sure how to do that (though maybe I'm misunderstanding your suggestion — I'm not too strong at Rust).

aaronArinder replied:
oh! try something like this `use blah::SuperLongName::BuildInput as ShorterBuildInput`

aaronArinder commented on docs/source/commands/supergraphs.mdx:
````diff
@@ -62,7 +62,7 @@ cat ./supergraph.yaml | rover supergraph compose --config -
 
 ```bash
-rover supergraph config --graph-ref platform@staging
+rover supergraph compose --graph-ref platform@staging
````
ty!

aaronArinder commented on src/utils/supergraph_config.rs:
```diff
@@ -47,7 +51,7 @@ impl RemoteSubgraphs {
             .into_iter()
             .map(|subgraph| (subgraph.name().clone(), subgraph.into()))
             .collect();
-        let supergraph_config = SupergraphConfig::new(subgraphs, federation_version.cloned());
+        let supergraph_config = SupergraphConfig::new(subgraphs, federation_version);
```
nice

aaronArinder commented on src/utils/supergraph_config.rs:
```diff
@@ -116,28 +119,35 @@ fn merge_supergraph_configs(
+        .or(remote_config
+            .as_ref()
+            .and_then(|it| it.get_federation_version()))
+        .unwrap_or(FederationVersion::LatestFedTwo);
```
probably want `.or_else` so that what's inside each `.or()` isn't evaluated (`.or_else` is lazily evaluated)

aaronArinder commented on src/utils/supergraph_config.rs:
```diff
@@ -116,28 +119,35 @@ fn merge_supergraph_configs(
 ) -> Option<SupergraphConfig> {
+    // We use the federation version explicitly given at the command line as top
+    // priority; otherwise the version explicitly given in the local config
+    // file; otherwise the version fetched from Studio.
```
I'd add something about defaulting to latest_fed_two just to make it explicit that if there's no target, no remote, and no local, we're going to use that one

aaronArinder reviewed (dismissed):
minor nits; only thing I really care about is adding some clarity to the comment about which federation version is in use

pre-approving, but definitely want the commented expanded with the final default

glasser committed:
compose/dev: don't delete remote routing_urls

This pulls in https://github.com/apollographql/federation-rs/pull/574
which improves the logic used to merge remote and local config such that
if your local config does not specify a `routing_url` for a subgraph
that also exists in the remote config, the remote config's URL is used
instead of deleting it.

Update the docs.

glasser committed:
Suggestions from code review

glasser committed:
More suggestions from code review

glasser committed:
Fix fmt

aaronArinder reviewed (approved):
ty for the changes! approving, but before merging, make sure the [smoke tests](https://github.com/apollographql/rover/actions/runs/10744012960) pass (running now but moving on to other things)
